API-Gateway module CR blocking deletion (#567)
kyma-project/api-gateway: docs/contributor/04-10-technical-design.md

The following diagram illustrates the APIRule reconciliation process and the resources created in the process:
![Kyma API Gateway Overview](../assets/operator-contributor-skr-overview.svg)

### APIGateway Controller

APIGateway Controller is a [Kubernetes controller](https://kubernetes.io/docs/concepts/architecture/controller/), which is implemented using the [Kubebuilder](https://book.kubebuilder.io/) framework.
The controller is responsible for handling the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md).

#### Reconciliation
The [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md) is reconciled with each change. If no changes have been made, the reconciliation process occurs at the default interval of 10 hours,
as determined by the [Kubernetes controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime).
If there is a failure during the reconciliation process, the default behavior of the [Kubernetes controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime) is to use exponential backoff requeue.

> On deletion of [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md), the controller will check before deleting it if there any APIRule or Istio [VirtualService](https://istio.io/latest/docs/reference/config/networking/virtual-service) resources which are referring to the Kyma default [Gateway](https://istio.io/latest/docs/reference/config/networking/gateway/) (`kyma-system/kyma-gateway`). If there are any found, they will be listed in the logs of the controller and [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md)'s status will be set to `Warning` explaining that there are resources blocking the deletion. Deletion of the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md) is protected with a finalizer named `gateways.operator.kyma-project.io/api-gateway-reconciliation`. Once there are no more APIRule and VirtualService resources blocking the deletion, the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md) will be deleted. Deletion of [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md) also deletes Kyma default Gateway.

Reviewer: ```suggestion
Before deleting the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md), the controller first checks if there are any APIRule or [Istio VirtualService](https://istio.io/latest/docs/reference/config/networking/virtual-service) resources that reference the default Kyma [Gateway](https://istio.io/latest/docs/reference/config/networking/gateway/) `kyma-system/kyma-gateway`. If any such resources are found, they are listed in the logs of the controller, and the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md)'s status is set to `Warning` to indicate that there are resources blocking the deletion. 
The `gateways.operator.kyma-project.io/api-gateway-reconciliation` finalizer protects the deletion of the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md). Once no more APIRule and VirtualService resources are blocking the deletion of the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md), the APIGateway CR can be deleted. Deleting the [APIGateway CR](../user/custom-resources/apigateway/04-00-apigateway-custom-resource.md) also deletes the default Kyma Gateway. 
```
